main/src/ca/uwaterloo/flix/api/Bootstrap.scala

```diff
@@ -28,59 +31,179 @@ import scala.collection.mutable
+  private var flixPackagePaths: List[Path] = List.empty
+  private var mavenPackagePaths: List[Path] = List.empty
+
+  def bootstrap(pathString: String): Result[List[Path], BootstrapError] = {
     //
     // Determine the mode: If `path/flix.toml` exists then "project" mode else "folder mode".
     //
-    if (???) {
-      projectMode()
+    val path = Paths.get(pathString)
+    val tomlPath = getManifestFile(path)
+    if (Files.exists(tomlPath)) {
+      projectMode(path)
     } else {
-      folderMode()
+      folderMode(path)
     }
   }
 
-  private def projectMode(): Unit = { // TODO: Probably return Result or Validation
+  private def projectMode(path: Path): Result[List[Path], BootstrapError] = {
     // 1. Read, parse, and validate flix.toml.
+    val tomlPath = getManifestFile(path)
+    val manifest = ManifestParser.parse(tomlPath) match {
+      case Ok(m) => m
+      case Err(e) => return Err(BootstrapError.ManifestParseError(e))
+    }
+
     // 2. Check each dependency is available or download it.
-    // 3. Compute the set of JAR paths and Flix fpkg paths.
-    // 4. Add *.flix, src/**.flix and test/**.flix
+    FlixPackageManager.installAll(manifest, path)(System.out) match {
+      case Ok(l) => flixPackagePaths = l
+      case Err(e) => return Err(BootstrapError.FlixPackageError(e))
+    }
+    MavenPackageManager.installAll(manifest)(System.out) match {
+      case Ok(l) => mavenPackagePaths = l
+      case Err(e) => return Err(BootstrapError.MavenPackageError(e))
+    }
+
+    // 3. Add *.flix, src/**.flix and test/**.flix
+    val filesHere = getAllFlixFilesHere(path)
+    val filesSrc = getAllFilesWithExt(getSourceDirectory(path), "flix")
+    val filesTest = getAllFilesWithExt(getTestDirectory(path), "flix")
+    sourcePaths = filesHere ++ filesSrc ++ filesTest
+
+    timestamps = timestamps ++ (flixPackagePaths ++ mavenPackagePaths ++ sourcePaths).map(f => f -> f.toFile.lastModified).toMap
+
+    Ok(sourcePaths ++ flixPackagePaths ++ mavenPackagePaths)
   }
 
-  private def folderMode(): Unit = { // TODO: Probably return Result or Validation
+  private def folderMode(path: Path): Result[List[Path], BootstrapError] = {
     // 1. Add *.flix, src/**.flix and test/**.flix
+    val filesHere = getAllFlixFilesHere(path)
+    val filesSrc = getAllFilesWithExt(getSourceDirectory(path), "flix")
+    val filesTest = getAllFilesWithExt(getTestDirectory(path), "flix")
+    sourcePaths = filesHere ++ filesSrc ++ filesTest
+
     // 2. Grab all jars in lib/
+    val jarFilesLib = getAllFilesWithExt(getLibraryDirectory(path), "jar")
+    mavenPackagePaths = jarFilesLib
+
     // 3. Grab all flix packages in lib/
+    val flixFilesLib = getAllFilesWithExt(getLibraryDirectory(path), "fpkg")
+    flixPackagePaths = flixFilesLib
+
+    Ok(sourcePaths ++ flixPackagePaths ++ mavenPackagePaths)
   }
 
   def reconfigureFlix(flix: Flix): Unit = { // TODO: Probably return Result or Validation
```

Comment:
@magnus-madsen can you take a look at this and let me know if I am on the right track?

Reply:
Should reconfigureFlix return a list of current paths?

Reply:
No, no need. It simply configures the Flix object.

Reply:
Then why should it not just return Unit? What could go wrong that I can detect?

Reply:
Yes, I guess just unit :)